fuzz/fuzz_targets/uncompress.rs

```diff
@@ -1,54 +1,147 @@
+//! Rust-specific UB, use Miri. Once a corpus with suitable coverage has been built, you can run
+//! Miri against the corpus by executing:
+//! ```
+//! MIRIFLAGS=-Zmiri-disable-isolation cargo miri nextest run --bin uncompress --features disable-checksum
+//! ```
+//! This assumes the corpus is located in the default directory of `corpus/uncompress`. If it
+//! isn't, specify the corpus directory with the `ZLIB_RS_CORPUS_DIR` environment variable.
+#![cfg_attr(not(any(miri, test)), no_main)]
 
-use zlib_rs::ReturnCode;
+use libfuzzer_sys::{fuzz_target, Corpus};
+use libz_rs_sys::{
+    gz_header, inflate, inflateEnd, inflateGetHeader, inflateInit2_, z_stream, zlibVersion,
+};
+use zlib_rs::{InflateFlush, ReturnCode};
 
-fn uncompress_help(input: &[u8]) -> Vec<u8> {
-    let mut dest_vec = vec![0u8; 1 << 16];
+fuzz_target!(|input: &[u8]| -> Corpus { run(input) });
 
-    let mut dest_len = dest_vec.len() as std::ffi::c_ulong;
-    let dest = dest_vec.as_mut_ptr();
+fn run(input: &[u8]) -> Corpus {
+    if input.is_empty() {
+        return Corpus::Reject;
+    }
 
-    let source = input.as_ptr();
-    let source_len = input.len() as _;
+    let mut stream = z_stream::default();
 
-    let err = unsafe { ::libz_rs_sys::uncompress(dest, &mut dest_len, source, source_len) };
+    let err = unsafe {
+        inflateInit2_(
+            &mut stream,
+            15 + 32, // Support both zlib and gzip files.
+            zlibVersion(),
+            size_of::<z_stream>() as _,
+        )
+    };
+    assert_eq!(ReturnCode::from(err), ReturnCode::Ok);
 
-    if err != 0 {
-        panic!("error {:?}", ReturnCode::from(err));
-    }
+    let mut extra = vec![0; 64];
+    let mut name = vec![0; 64];
+    let mut comment = vec![0; 64];
+    let mut header = gz_header {
+        text: 0,
+        time: 0,
+        xflags: 0,
+        os: 0,
+        extra: extra.as_mut_ptr(),
+        extra_len: 0,
+        extra_max: 64,
+        name: name.as_mut_ptr(),
+        name_max: 64,
+        comment: comment.as_mut_ptr(),
+        comm_max: 64,
+        hcrc: 0,
+        done: 0,
+    };
 
-    dest_vec.truncate(dest_len as usize);
+    let err = unsafe { inflateGetHeader(&mut stream, &mut header) };
+    assert_eq!(ReturnCode::from(err), ReturnCode::Ok);
 
-    dest_vec
-}
+    let mut output = vec![0; input.len()];
+    let input_len: u64 = input.len().try_into().unwrap();
+    stream.next_out = output.as_mut_ptr();
+    stream.avail_out = output.len().try_into().unwrap();
 
-fuzz_target!(|data: String| {
-    // first, deflate the data using the standard zlib
-    let mut length = 8 * 1024;
-    let mut deflated = vec![0; length];
-
-    let error = unsafe {
-        libz_ng_sys::compress(
-            deflated.as_mut_ptr().cast(),
-            &mut length,
-            data.as_ptr().cast(),
-            data.len(),
-        )
+    // Small enough to hit interesting cases, but large enough to hit the fast path
+    let chunk_size = 64;
```

Comment:
Would varying the chunk size make sense? Especially non-power-of-two chunk sizes seem like they could be useful for finding edge-cases.

Reply:
It might, but I don't see how we could do that in a deterministic way and continue to use the compression corpus.

Reply:
You could take the chunk size from the fuzz input, right? So for example interpreting the fuzz input as series of 1 byte chunk size + n bytes chunk data.

Reply:
Would that not mess with how effective the corpus is?

Reply:
You could reformat the existing corpus into this format, right? In any case I guess this is better left as future improvement.